### PR TITLE
blob storage API for local storage and google cloud storage #681

### PR DESCRIPTION
**Pull request by alexis-kochetov**

### CODE REVIEW

File: src/dotnet/Core/Blobs/ByteStreamExt.cs [hunk 1]
```diff
@@ -61,7 +60,7 @@ await foreach (var blobPart in byteStream.ConfigureAwait(false))
 
         var bytesWritten = await stream.WriteByteStream(byteStream, false, cancellationToken).ConfigureAwait(false);
         stream.Position = 0;
-        await target.WriteAsync(blobId, stream, false, cancellationToken).ConfigureAwait(false);
+        await target.Write(blobId, stream, "application/octet-stream", cancellationToken).ConfigureAwait(false);
```
Comment (crui3er) [inline src/dotnet/Core/Blobs/ByteStreamExt.cs:63]: Better replace with MediaTypeNames.Application.Octet

File: src/dotnet/Core/Blobs/Internal/LocalFolderBlobStorage.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        }
+    }
+
+    public Task<string?> GetContentType(string path, CancellationToken cancellationToken)
+    {
+        if (path == null) throw new ArgumentNullException(nameof(path));
+
+        var filePath = FilePath.New(path);
+        if (!filePath.IsSubPathOf(_directory) && filePath.IsRooted)
+            throw StandardError.Constraint<LocalFolderBlobStorage>("Path should be relative to the .ctor directory.");
```
Comment (crui3er) [inline src/dotnet/Core/Blobs/Internal/LocalFolderBlobStorage.cs:43]: What does '.ctor directory' mean?

File: src/dotnet/Core/Blobs/Internal/GoogleCloudBlobStorage.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        _bucket = bucket;
+        _client = StorageClient.Create();
+    }
+
+    public async Task<Stream?> Read(string path, CancellationToken cancellationToken)
+    {
+        if (path == null) throw new ArgumentNullException(nameof(path));
+
+        var stream = new MemoryStream();
```
Comment (crui3er) [inline src/dotnet/Core/Blobs/Internal/GoogleCloudBlobStorage.cs:22]: Can we also use here recycleable memory streams? Or does they not suit for this task?

File: src/dotnet/Core/Blobs/Internal/LocalFolderBlobStorage.cs (hunk 2 above)
Comment (alexis-kochetov) [reply src/dotnet/Core/Blobs/Internal/LocalFolderBlobStorage.cs:43]: probably I should rename to base

File: src/dotnet/Core/Blobs/Internal/GoogleCloudBlobStorage.cs (hunk 3 above)
Comment (alexis-kochetov) [reply src/dotnet/Core/Blobs/Internal/GoogleCloudBlobStorage.cs:22]: Let me check

File: src/dotnet/Core/Blobs/Internal/GoogleCloudBlobStorage.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    public async Task Delete(IReadOnlyCollection<string> paths, CancellationToken cancellationToken)
+    {
+        if (paths == null) throw new ArgumentNullException(nameof(paths));
+        if (paths.Count == 0)
+            return;
+
+        var resultTasks = new List<Task>(paths.Count);
+        resultTasks.AddRange(paths.Select(path => _client.DeleteObjectAsync(_bucket, path, cancellationToken: cancellationToken)));
```
Comment (FrolYo) [inline src/dotnet/Core/Blobs/Internal/GoogleCloudBlobStorage.cs:68]: just a suggestion
```CSharp
await paths.Select(path => _client.DeleteObjectAsync(_bucket, path, cancellationToken)).Collect().ConfigureAwait(false)
```

**Review by FrolYo — APPROVED** [review]

**Pull request merged**